runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/graph/GreedyPipelineFuserTest.java

```diff
@@ -0,0 +1,857 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.core.construction.graph;
+
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.hasSize;
+import static org.junit.Assert.assertThat;
+
+import org.apache.beam.model.pipeline.v1.RunnerApi.Components;
+import org.apache.beam.model.pipeline.v1.RunnerApi.Environment;
+import org.apache.beam.model.pipeline.v1.RunnerApi.FunctionSpec;
+import org.apache.beam.model.pipeline.v1.RunnerApi.PCollection;
+import org.apache.beam.model.pipeline.v1.RunnerApi.PTransform;
+import org.apache.beam.model.pipeline.v1.RunnerApi.ParDoPayload;
+import org.apache.beam.model.pipeline.v1.RunnerApi.Pipeline;
+import org.apache.beam.model.pipeline.v1.RunnerApi.SdkFunctionSpec;
+import org.apache.beam.model.pipeline.v1.RunnerApi.SideInput;
+import org.apache.beam.model.pipeline.v1.RunnerApi.WindowIntoPayload;
+import org.apache.beam.runners.core.construction.PTransformTranslation;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Tests for {@link GreedyPipelineFuser}. */
+@RunWith(JUnit4.class)
+public class GreedyPipelineFuserTest {
+  // Contains the 'go' and 'py' environments, and a default 'impulse' step and output.
+  private Components partialComponents;
+
+  @Before
+  public void setup() {
+    partialComponents =
+        Components.newBuilder()
+            .putTransforms(
+                "impulse",
+                PTransform.newBuilder()
+                    .putOutputs("output", "impulse.out")
+                    .setSpec(
+                        FunctionSpec.newBuilder()
+                            .setUrn(PTransformTranslation.IMPULSE_TRANSFORM_URN))
+                    .build())
+            .putPcollections(
+                "impulse.out", PCollection.newBuilder().setUniqueName("impulse.out").build())
+            .putEnvironments("go", Environment.newBuilder().setUrl("go").build())
+            .putEnvironments("py", Environment.newBuilder().setUrl("py").build())
+            .build();
+  }
+
+  /*
+   * impulse -> .out -> read -> .out -> parDo -> .out -> window -> .out
+   * becomes
+   * (impulse.out) -> read -> read.out -> parDo -> parDo.out -> window
```

Comment:
window.out should appear inside the fused stage, ditto on any leaf nodes below where the PCollection does not need to be materialized.

Reply:
It exists within the stage, but it's not one of the outputs because it has no external consumers (the `PCollection`, but not its producer, is "fused away")